h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -13,7 +15,7 @@
 
 public interface HyperSpaceWalker<MP extends Model.Parameters, C extends HyperSpaceSearchCriteria> {
 
-  interface HyperSpaceIterator<MP extends Model.Parameters> {
+  abstract class HyperSpaceIterator<MP extends Model.Parameters> {
```

Comment:
this is probably a breaking API change and another reason to implement this as a wrapper over iterator

Reply:
@honzasterba could you please point out to the place where this change is breaking API? I might be missing something. 
From the high-level perspective interface is special case of the abstract class, I would say. Interface just don't have default behaviour, by default have public modifiers etc ( see for example here https://javapapers.com/core-java/abstract-and-interface-core-java-2/difference-between-a-java-interface-and-a-java-abstract-class/ ). The only thing that I would concern about is that me can implement many interfaces but we can extend only one abstract class.

We actually do have similar situation with walkers themselves - we have `BaseWalker` which is an abstract class. I can consider to leave `HyperSpaceIterator` as an interface and provide abstract class that implements this interface with additional default methods(like `BaseWalker` do). But I'm not sure whether it makes sense....all walkers actually extends only `BaseWalker`, and only `BaseWalker` implements `HyperSpaceWalker` interface.
@michalkurka could you please help us to reason about legacy/existing design solutions and current one?

Reply:
someone might be implementing this interface, by converting this to a class it would break their code

Reply:
@honzasterba fair point. My point was that it is a synthetic interface. It has too many things inside(max_models, time remaining, model's failures) and should have been an abstract class or be separated in multiple interfaces(Limitable, Timable,Failurable as an example). Interface usually cover one aspect of something. I doubt somebody is implementing it but we never know - will fix this. Thanks!